### PR TITLE
drivers/bmx280: reworked driver and added SPI mode

### PR DESCRIPTION
**Pull request by haukepetersen**
### Issues/PRs references
For supporting  the `ruuvitag` it is needed to interface the `bmx280` sensor in SPI mode. While adding this mode, I found the driver was in a very bad state. This PR adds the possibility to use the driver in SPI mode and it significantly improves the drivers structure and style:

 bugs fixed (I wonder how these were able to go by previous reviews...):
- move global variables into device descriptor
- guard bus access (use acquire and release)

added functionality:
- enable SPI mode

structural improvements:
- reduce stack usage
- simplify the driver's structure
- centralize bus access code
- use assertions
- cleanup includes
- use shortcuts for bus access

style changes:
- fix line length
- cleanup and improve doxygen
- unify pointer notation (char *var over char* var)
- unify (error) return messages
- use `#ifdef MODULE_BME280` instead of `#if defined(BME..)`
- unify debug messages -> using `[bmx28] x: msg` scheme

This PR also adapts and improves the test application accordingly and also add the `bme280_spi` module as default SAUL module to the `ruuvitag`.

Note on commits: as it was impossible to cleanly split the rework and cleanup efforts from the SPI mode addition to the driver, I put all changes into a single commit...

### Testing:
The changes are tested and verified for the `bmx280_spi` and `bmp280_spi` on the `ruuvitag`. I don't have access to any board having these sensors connected via I2C, so the I2C mode still needs to be tested.

### CODE REVIEW

**Comment by haukepetersen** [comment]
rebased and fixed initialization and init_master order for the i2c mode (according to #8446).

**Comment by haukepetersen** [comment]
rebased. Anyone willing to review?

File: drivers/Makefile.dep [hunk 1]
```diff
@@ -47,14 +47,27 @@ ifneq (,$(filter bmp180,$(USEMODULE)))
   USEMODULE += xtimer
 endif
 
-ifneq (,$(filter bmx055,$(USEMODULE)))
-  FEATURES_REQUIRED += periph_i2c
+ifneq (,$(filter bme280_spi,$(USEMODULE)))
```
Comment (aabadie) [inline drivers/Makefile.dep]: Why not use the wildcard `%` and merge this one with the block under ? I think the wildcard could also be used in the block below that handles the distinction between spi and i2c.

File: drivers/bmx280/bmx280.c [hunk 2]
```diff
@@ -23,99 +25,329 @@
...
+
+    /* write basic device configuration: t_sb and filter values */
+    reg = (dev->params.t_sb | dev->params.filter);
+    if (_write_reg(dev, BMX280_CONFIG_REG, reg) != BMX280_OK) {
+        goto err;
+    }
+
+#ifdef MODULE_BME280
+    /* ctrl_hum must be written before ctrl_meas for changes to become
+     * effective ... */
```
Comment (aabadie) [inline drivers/bmx280/bmx280.c]: The 3 dots are superfluous IMHO

File: tests/driver_bmx280/main.c [hunk 3]
```diff
@@ -1,32 +1,34 @@
...
 #include <stdio.h>
-#include <inttypes.h>
 
 #include "bmx280_params.h"
 #include "bmx280.h"
 #include "xtimer.h"
+#include "fmt.h"
 
-#define MAINLOOP_DELAY  (2 * 1000 * 1000u)      /* 2 seconds delay between printf's */
+#define MAINLOOP_DELAY  (2 * US_PER_SEC)      /* read sensor every 2 seconds */
```
Comment (aabadie) [inline tests/driver_bmx280/main.c]: We could just use xtimer_sleep(2) below and drop this line I think

**Comment by aabadie** [comment]
Maybe @dylad is interested by this.

**Comment by haukepetersen** [comment]
Sorry for the subjective grading, there was a high level of frustration involved when working on this driver so I got in a bad mood...

I know that splitting this PR would make a lot of sense from a reviewing perspective, but unfortunately this is not so easy. Adding the new SPI feature and reworking the driver's structure is highly correlated, so that splitting the changes would probably take another 1-2 days on top, which is not feasible for me to do. Do you guys think we could get this in somehow in the current state?

**Comment by aabadie** [comment]
> Do you guys think we could get this in somehow in the current state?

Yes, on my side. But we need to wait for the I2C embargo first.

**Comment by haukepetersen** [comment]
awesome, thanks! Then lets get back to this PR once the I2C stuff is merged.

File: tests/driver_bmx280/main.c [hunk 4]
```diff
@@ -35,21 +37,27 @@ int main(void)
...
-    }
+    if (result != BMX280_OK) {
+        switch (result) {
+            case BMX280_ERR_BUS:
+                puts("[Error] The given i2c is not enabled");
+                break;
+            case BMX280_ERR_NODEV:
+                puts("[Error] Unable to communicate with any BMX280 device");
+                break;
+            default:
```
Comment (nmeum) [inline tests/driver_bmx280/main.c]: You forgot about `BMX280_ERR_NOCAL`. See #9319

**Comment by PeterKietzmann** [comment]
> awesome, thanks! Then lets get back to this PR once the I2C stuff is merged.

It is!

**Comment by haukepetersen** [comment]
Will try to find the time and rebase this during the next days, stay tuned :-) (and sorry for the delay!)

**Comment by haukepetersen** [comment]
finally adapted and fixed this driver to work with the current master.

Tested as before on `ruuvitag` (SPI mode) but not for the `sltb001a` (I2C mode)

**Comment by haukepetersen** [comment]
rebased once more.

@nmeum @aabadie @PeterKietzmann would anyone mind to take another look?! Thanks.

File: drivers/bmx280/bmx280.c [hunk 5]
```diff
@@ -23,93 +25,320 @@
...
+
+static int16_t _to_i16_le(const uint8_t *buffer, size_t offset)
+{
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+    return (((int16_t)buffer[offset + 1]) << 8) + buffer[offset];
+#else
+    return (((int16_t)buffer[offset]) << 8) + buffer[offset + 1];
+#endif
+}
+
```
Comment (benpicco) [inline drivers/bmx280/bmx280.c]: We should have this in common code.

File: drivers/bmx280/bmx280.c [hunk 6]
```diff
@@ -23,93 +25,320 @@
...
 #include "log.h"
+#include "assert.h"
 #include "bmx280.h"
 #include "bmx280_internals.h"
-#include "bmx280_params.h"
-#include "periph/i2c.h"
 #include "xtimer.h"
 
-#define ENABLE_DEBUG        (0)
+#define ENABLE_DEBUG        (1)
```
Comment (benpicco) [inline drivers/bmx280/bmx280.c]: Better leave that at `(0)` :wink:

**Comment by benpicco** [comment]
The driver looks good, you'll need to update `
tests/driver_bmx280/Makefile.ci` though.

File: drivers/bmx280/bmx280.c (hunk 5 above)
Comment (haukepetersen) [reply drivers/bmx280/bmx280.c]: of course, its long been merged into byteorder.h -> a nice sign of how old this PR has become :-) Will fix.

File: drivers/bmx280/bmx280.c (hunk 6 above)
Comment (haukepetersen) [reply drivers/bmx280/bmx280.c]: of course

File: drivers/bmx280/bmx280.c (hunk 5 above)
Comment (haukepetersen) [reply drivers/bmx280/bmx280.c]: ups, on closer look it seems I was mistaken. These functions do not exist in shared code anywhere until now. IMO they are borderline in terms of being generic enough to be moved into `byteorder.h` or similar or just stay in this module. I'd like to merge them as is to take some more time to think about what better to do here...

File: drivers/bmx280/bmx280.c [hunk 7]
```diff
@@ -23,93 +25,320 @@
...
-static uint16_t get_uint16_le(const uint8_t *buffer, size_t offset);
-static int16_t get_int16_le(const uint8_t *buffer, size_t offset);
-
-#if ENABLE_DEBUG
-static void dump_buffer(const char *txt, uint8_t *buffer, size_t size);
-#define DUMP_BUFFER(txt, buffer, size)  dump_buffer(txt, buffer, size)
+#ifdef BMX280_USE_SPI
+#define BUS                 (dev->params.spi)
+#define CS                  (dev->params.cs)
+#define CLK                 SPI_CLK_5MHZ
```
Comment (benpicco) [inline drivers/bmx280/bmx280.c]: I think this should go to `bmx280_params.h`

File: drivers/bmx280/bmx280.c [hunk 8]
```diff
@@ -23,93 +25,320 @@
...
- * These are read all at once and then used to compute the three sensor values.
+ * @brief   Trigger a new measurement (if applicable) and read raw data
  */
-static uint8_t measurement_regs[8];
+static int _do_measurement(bmx280_t *dev)
+{
+    uint8_t reg;
+
+    /* get access to the bus */
+    _acquire(dev);
```
Comment (benpicco) [inline drivers/bmx280/bmx280.c]: No error checking?

File: drivers/bmx280/bmx280.c (hunk 5 above)
Comment (benpicco) [reply drivers/bmx280/bmx280.c]: Functions to read from an unaligned  buffer are always useful.

File: drivers/bmx280/bmx280.c [hunk 9]
```diff
@@ -166,218 +394,39 @@ uint32_t bmx280_read_pressure(const bmx280_t *dev)
...
     var1 = (((((adc_H << 14) - (((int32_t)cal->dig_H4) << 20) - (((int32_t)cal->dig_H5) * var1)) +
               ((int32_t)16384)) >> 15) *
             (((((((var1 * ((int32_t)cal->dig_H6)) >> 10) *
                  (((var1 * ((int32_t)cal->dig_H3)) >> 11) + ((int32_t)32768))) >> 10) +
                ((int32_t)2097152)) * ((int32_t)cal->dig_H2) + 8192) >> 14));
     var1 = (var1 - (((((var1 >> 15) * (var1 >> 15)) >> 7) * ((int32_t)cal->dig_H1)) >> 4));
-    var1 = (var1 < 0 ? 0 : var1);
-    var1 = (var1 > 419430400 ? 419430400 : var1);
+    var1 = (var1 < 0) ? 0 : var1;
+    var1 = (var1 > 419430400) ? 419430400 : var1;
```
Comment (benpicco) [inline drivers/bmx280/bmx280.c]: You don't happen to know the source of those magic values?

File: drivers/bmx280/bmx280.c (hunk 7 above)
Comment (haukepetersen) [reply drivers/bmx280/bmx280.c]: Makes sense -> fixed.

File: drivers/bmx280/bmx280.c (hunk 8 above)
Comment (haukepetersen) [reply drivers/bmx280/bmx280.c]: fixed.

File: drivers/bmx280/bmx280.c (hunk 9 above)
Comment (haukepetersen) [reply drivers/bmx280/bmx280.c]: nope. This code was already part of the driver before I touched it. It is magic code that is given by Bosch if I am not mistaken...

**Comment by haukepetersen** [comment]
fixed latest findings, let me know when I should squash.

**Comment by benpicco** [comment]
`avr-rss2` will need an update too.
You can squash while you're at it.

**Comment by haukepetersen** [comment]
done

File: tests/driver_bmx280/Makefile.ci [hunk 10]
```diff
@@ -0,0 +1,7 @@
+BOARD_INSUFFICIENT_MEMORY := \
+    arduino-duemilanove \
+    arduino-nano \
+    arduino-uno \
+    atmega328p \
+    avr-rss2 \
```
Comment (benpicco) [inline tests/driver_bmx280/Makefile.ci]: I meant the [`Makefile.dep`](https://github.com/RIOT-OS/RIOT/blob/master/boards/avr-rss2/Makefile.dep) need an update.
The board should have enough memory.

Comment (haukepetersen) [reply tests/driver_bmx280/Makefile.ci]: of course, sorry. Was in auto-mode this morning...

**Comment by haukepetersen** [comment]
now the `avr-rss2` should be happy for real :-)

**Comment by benpicco** [comment]
Maybe now @herjulf could check out this branch and confirm that the I2C mode of the driver still works after the rewrite 😉

**Comment by herjulf** [comment]
bme280 on i2c using avr-rss2 tests/driver_bmx280
-------------------------------------------------------------------
> 
+--------Starting Measurements--------+
Temperature [°C]: 21.7
Pressure [Pa]: 102318
Humidity [%rH]: 42.01

+-------------------------------------+
Temperature [°C]: 21.7
Pressure [Pa]: 102309
Humidity [%rH]: 41.98

with patch
---------------

+--------Starting Measurements--------+
Temperature [°C]: 21.7
Pressure [Pa]: 102319
Humidity [%rH]: 42.13

+-------------------------------------+
Temperature [°C]: 21.7
Pressure [Pa]: 102314
Humidity [%rH]: 42.09

looks good
---------------

**Comment by herjulf** [comment]
Huh. Test was not good
--------------------------------
Please fix:

> main(): This is RIOT! (Version: 2020.01-devel-767-g8836f-at24mac)

BMX280 test application

+------------Initializing------------+
[Error] Unable to communicate with any BMX280 device

File: drivers/Makefile.dep [hunk 11]
```diff
@@ -75,19 +75,32 @@ ifneq (,$(filter bmp180,$(USEMODULE)))
...
+ifneq (,$(filter bmx055,$(USEMODULE)))
+  FEATURES_REQUIRED += periph_i2c
+endif
+
 ifneq (,$(filter cc110%,$(USEMODULE)))
   USEMODULE += cc110x
   USEMODULE += cc1xxx_common
+endif
+
+ifneq (,$(filter cc110x,$(USEMODULE)))
```
Comment (aabadie) [inline drivers/Makefile.dep]: unrelated change ?

File: tests/driver_bmx280/main.c [hunk 12]
```diff
@@ -36,21 +38,23 @@ int main(void)
 
     puts("BMX280 test application\n");
 
-    printf("+------------Initializing------------+\n");
+    puts("+------------Initializing------------+");
     result = bmx280_init(&dev, &bmx280_params[0]);
-    if (result == -1) {
-        puts("[Error] The given i2c is not enabled");
-        return 1;
+    switch (result) {
```
Comment (aabadie) [inline tests/driver_bmx280/main.c]: Nit:

```suggestion
    switch (bmx280_init(&dev, &bmx280_params[0])) {
```

and the result variable can be removed.

**Comment by haukepetersen** [comment]
@aabadie are you ok with this PR? Would be nice to cross it of the list after all this time... And as already promised, next time I will approach this type of fix more gentle for easier review ;-)

File: drivers/Makefile.dep (hunk 11 above)
Comment (haukepetersen) [reply drivers/Makefile.dep]: yes, seems to be a broken rebase

**Comment by haukepetersen** [comment]
never mind, seems like we wrote our comments in the same 3 minutes... :-)

**Comment by haukepetersen** [comment]
@herjulf

Could you give slightly more context? Which Board? Which driver config (bmp/bme, spi/i2c)?

**Comment by haukepetersen** [comment]
@aabadie your comments are addressed.

**Comment by herjulf** [comment]
@haukepetersen 
Problem with BME280 on i2c. avr-rss2 board.

**Comment by haukepetersen** [comment]
Unfortunately I don't have any board with an `bmx280` in I2C configuration available, so I am not able to test/debug that setup.

But I did just verify the SPI mode on the `ruuvitag` one more:
```2019-11-22 10:54:13,807 # main(): This is RIOT! (Version: 2020.01-devel-824-g71267-opt_driver_bmx280spi)
2019-11-22 10:54:13,808 # BMX280 test application
2019-11-22 10:54:13,808 # 
2019-11-22 10:54:13,808 # +------------Initializing------------+
2019-11-22 10:54:13,809 # Initialization successful
2019-11-22 10:54:13,809 # 
2019-11-22 10:54:13,809 # +------------Calibration Data------------+
2019-11-22 10:54:13,810 # dig_T1: 28059
2019-11-22 10:54:13,810 # dig_T2: 26414
2019-11-22 10:54:13,810 # dig_T3: 50
2019-11-22 10:54:13,810 # dig_P1: 38795
2019-11-22 10:54:13,811 # dig_P2: -10696
2019-11-22 10:54:13,811 # dig_P3: 3024
2019-11-22 10:54:13,811 # dig_P4: 8503
2019-11-22 10:54:13,811 # dig_P5: -287
2019-11-22 10:54:13,811 # dig_P6: 13049
2019-11-22 10:54:13,812 # dig_P7: 0
2019-11-22 10:54:13,812 # dig_P8: 7
2019-11-22 10:54:13,812 # dig_P9: 0
2019-11-22 10:54:13,812 # dig_H1: 0
2019-11-22 10:54:13,813 # dig_H2: 380
2019-11-22 10:54:13,813 # dig_H3: 0
2019-11-22 10:54:13,813 # dig_H4: 277
2019-11-22 10:54:13,813 # dig_H5: 50
2019-11-22 10:54:13,813 # dig_H6: 30
2019-11-22 10:54:13,814 # 
2019-11-22 10:54:13,814 # +--------Starting Measurements--------+
2019-11-22 10:54:13,814 # Temperature [°C]: 26.16
2019-11-22 10:54:13,815 #    Pressure [PaTemperature [°C]: 26.17
2019-11-22 10:54:13,815 #    Pressure [Pa]: 100430
2019-11-22 10:54:13,816 #   Humidity [%rH]: 38.81
2019-11-22 10:54:13,816 # 
2019-11-22 10:54:13,816 # +-------------------------------------+
2019-11-22 10:54:13,817 # 
2019-11-22 10:54:15,457 # Temperature [°C]: 26.18
2019-11-22 10:54:15,457 #    Pressure [Pa]: 100426
2019-11-22 10:54:15,458 #   Humidity [%rH]: 38.89
2019-11-22 10:54:15,458 # 
2019-11-22 10:54:15,458 # +-------------------------------------+
2019-11-22 10:54:15,458 # 
2019-11-22 10:54:17,453 # Temperature [°C]: 26.18
2019-11-22 10:54:17,454 #    Pressure [Pa]: 100436
2019-11-22 10:54:17,454 #   Humidity [%rH]: 37.22
2019-11-22 10:54:17,454 # 
2019-11-22 10:54:17,455 # +-------------------------------------+
2019-11-22 10:54:17,455 # 
2019-11-22 10:54:19,446 # Temperature [°C]: 26.18
2019-11-22 10:54:19,447 #    Pressure [Pa]: 100431
2019-11-22 10:54:19,447 #   Humidity [%rH]: 36.21
2019-11-22 10:54:19,447 # 
2019-11-22 10:54:19,447 # +-------------------------------------+
2019-11-22 10:54:19,447 # 
2019-11-22 10:54:21,456 # Temperature [°C]: 26.19
2019-11-22 10:54:21,457 #    Pressure [Pa]: 100429
2019-11-22 10:54:21,457 #   Humidity [%rH]: 35.92
2019-11-22 10:54:21,457 # 
2019-11-22 10:54:21,458 # +-------------------------------------+
```

@herjulf do you have any idea what could have happened to the I2C on your board?

**Comment by herjulf** [comment]
Very first accesses reading the chip_id. Something missing in the i2c setup.
Not in office now.

**Comment by maribu** [comment]
Don' t we always require that all driver can work with multiple devices? The current architecture prevents using I2C attached and SPI attached BMx280 on the same board. IMO this conflicts with our driver guidelines.

Why not providing the common code as base module and the SPI and I2C backends as submodule? That would allow using both modes and gets rid of a lot of `#if`s and `#ifdef`s. And if only one flavour is used, the other submodule is not compiled in.

File: drivers/bmx280/bmx280.c [hunk 13]
```diff
@@ -23,93 +25,322 @@
...
+}
+
+static inline void _release(const bmx280_t *dev)
+{
+    i2c_release(BUS);
+}
+
+static int _read_reg(const bmx280_t *dev, uint8_t reg, uint8_t *data)
+{
+    if (i2c_read_reg(BUS, ADDR, reg, &data, 0) != 1) {
```
Comment (daexel) [inline drivers/bmx280/bmx280.c]: I tested the bme280 driver. An I got this errormessage: [bmx280] error: unable to read chip ID from device 
So you you have to change the _read_reg method. data is already a pointer and  the condition have to be 0.

**Comment by haukepetersen** [comment]
@daexel thanks for the pointer, very valid point!

I always thought that I did these adaptions to the 'new' i2c interface a long time ago - either I fucked up rebasing or I just did not do it in the first place... Sorry for that!

**Comment by haukepetersen** [comment]
> Don' t we always require that all driver can work with multiple devices? The current architecture prevents using I2C attached and SPI attached BMx280 on the same board. IMO this conflicts with our driver guidelines.

Yes and no. Support for multiple devices (if applicable) yes, but support for different driver modes simultaneously: definitively NO.

**Comment by herjulf** [comment]
Better..                                                                                                          
> Temperature [�°C]: 20.31                                                                                                    
>    Pressure [Pa]: 103965                                                                                                    
>   Humidity [%rH]: 31.78
>

**Comment by benpicco** [comment]
@haukepetersen are you aware of #12772? 
If it's a simple fix, it's probably better to just include it here to avoid the additional churn.

**Comment by haukepetersen** [comment]
Nope, I was not aware of #12772. But IMHO we might just handle both PRs separately, it should be straight forward to rebase #12772 on top of this PR, right?

**Comment by haukepetersen** [comment]
@herjulf thanks! I took the freedom to squash again...

**Comment by benpicco** [comment]
Yea but I thought when you are editing the file anyway it would be trivial to add those lines.
`_do_measurement()` is only called in `bmx280_read_temperature()` and since both other functions rely on the temperature too, it makes sense to just call it there.

**Comment by haukepetersen** [comment]
But then I get in trouble for not keeping authorship for selected contributions :-)

**Comment by haukepetersen** [comment]
> I don't want this to go in without a proper discussion about the architecture of the driver. I'll add reasons later.

Are you serious? Please open a separate issue, you are always welcome to rework this driver, but please lets get it merged for now!

**Comment by maribu** [comment]
@haukepetersen: Out of all people working on RIOT, you're the single person that really just should not complain if a PR is blocked last minute.

**Comment by benpicco** [comment]
I also think that an iterative approach is sensible here.
Just because the code is merged doesn't mean it's forever frozen in ice.
I agree that a modular architecture is more desirable, but this is not the only driver that uses this 'old' pattern. I think having a working state committed and then improving on that has more values for users of the SPI version now than optimizing for people who might want to connect both the I2C and the SPI version of that sensor to the same board.

**Comment by haukepetersen** [comment]
Its not about the last minute, its about the reasoning. The architecture of this driver as proposed in this PR is in line with the driver design as of now. I think we can all agree, that this PR makes the situation better than it was before the PR, so it does not hurt. Also it does not introduce anything architecturally weird that breaks with design practices, so again, IMO no blocker there.

Now starting a more generic discussion on how to design drivers in general is something we can always have, but please in a generic context!

File: drivers/bmx280/bmx280.c [hunk 14]
```diff
@@ -23,93 +25,322 @@
...
+    }
+
+#if defined(MODULE_BME280_SPI) || defined(MODULE_BME280_I2C)
+    /* ctrl_hum must be written before ctrl_meas for changes to become
+     * effective */
+    reg = dev->params.humid_oversample;
+    if (_write_reg(dev, BME280_CTRL_HUM_REG, reg) != BMX280_OK) {
+        goto err;
+    }
+#endif
```
Comment (maribu) [inline drivers/bmx280/bmx280.c]: It is not even possible to have a BMP280 and a BME280 using the same interface.

**Comment by kaspar030** [comment]
> Now starting a more generic discussion on how to design drivers in general is something we can always have, but please in a generic context!

+1.

@benpicco #12772 is a distinct change and should be seperate, if it eases reviewing. This PR's review is basically done, so adding changes will certainly not ease reviewing. Actually, at this point it is against procedures. Also, the other change is from a different author and came months after this PR. Let's not force re-opening this PR on @haukepetersen, integrate #12772 mentally, ...

@maribu IIRC this PR doesn't change the driver's architecture by much, but improves the current implementation. The architecture discussion should be kept seperate.

**Comment by maribu** [comment]
> but support for different driver modes simultaneously: definitively NO.

It was a deliberate design choice to implement bus support as different driver modes. And this design choice is what I challenge.

> Now starting a more generic discussion on how to design drivers in general is something we can always have, but please in a generic context!

It is not really my intent to start a discussion, but rather enforce general design goals already agreed upon.

Citing [RIOT's driver guide](https://github.com/RIOT-OS/RIOT/wiki/Guide%3A-Writing-a-device-driver-in-RIOT):

> Third, it should always be possible, to handle more than a single device of one kind. Drivers and their interfaces are thus designed to keep their state information in a parameterized location instead of driver defined global variables.

The wording of the sentence is not super clear, but to me this clearly means that a driver should not prevent a user from using multiple devices of the same kind even though attached to different buses.

Also there is a broad consensus on the use if `ifdef`s: They should be avoided, unless solid technical reasons can be found to justify them. (See e.g. https://github.com/RIOT-OS/RIOT/pull/12626 as a recent confirmation of that consensus.)

Not properly factoring out yields to such cases. This is clearly not the worst, e.g. when looking at the `at86rf2xx` driver we in stuff like this:

https://github.com/RIOT-OS/RIOT/blob/master/drivers/at86rf2xx/at86rf2xx.c#L51-58

https://github.com/RIOT-OS/RIOT/blob/master/drivers/at86rf2xx/at86rf2xx.c#L63-L71

https://github.com/RIOT-OS/RIOT/blob/master/drivers/at86rf2xx/at86rf2xx.c#L76-L87

https://github.com/RIOT-OS/RIOT/blob/master/drivers/at86rf2xx/at86rf2xx.c#L127-L136

https://github.com/RIOT-OS/RIOT/blob/master/drivers/at86rf2xx/at86rf2xx_getset.c#L232-L240

https://github.com/RIOT-OS/RIOT/blob/master/drivers/at86rf2xx/at86rf2xx_internal.c#L147-L154

https://github.com/RIOT-OS/RIOT/blob/master/drivers/at86rf2xx/at86rf2xx_netdev.c#L218-L226

Someone starts to go the lazy route to just hack in "flavors" of the driver using `#if`s and `#ifdef`s, and next time you like they are all over the place.

File: drivers/bmx280/bmx280.c (hunk 14 above)
Comment (haukepetersen) [reply drivers/bmx280/bmx280.c]: Nope, it is not. BUT
a) it is an extremely rare use case to have both connected concurrently 
b) you might just use the `bme280` driver config with both devices as a workaround
c) the convention in RIOT we always used to follow is not to over complicate things for the last 1% of exotic use cases
d) **this has nothing to do with the content of this PR**

**Comment by maribu** [comment]
> I also think that an iterative approach is sensible here.
> Just because the code is merged doesn't mean it's forever frozen in ice.

Agreed. And maybe it makes sense to migrate the driver guideline to github, so this is more visible and content changes in there are properly reviewed.

**Comment by haukepetersen** [comment]
> Citing RIOT's driver guide:

>> Third, it should always be possible, to handle more than a single device of one kind. Drivers and their interfaces are thus designed to keep their state information in a parameterized location instead of driver defined global variables.

It happens that I know that paragraph quite well :-) In general the state of the `bmx280` driver does actually comply to that paragraph - at least to me the `of one kind` means that we should support e.g. multiple `bme280` or multiple `bmp280`, but mixing them is IMHO not `of one kind`.

But don't get me wrong, I am not against discussing and adapting the design guide lines and best practices in driver design for RIOT - after all circumstances change, we all gain experience, and hence we assess decisions differently now. I think we should just keep those generic discussions should not per default block everything in its path...

**Comment by haukepetersen** [comment]
@benpicco all green, would you do me the honor?

@maribu thanks for yielding in this case. Looking forward to the design discussion. Especially how a fully dynamic configuration will compare in terms of efficiency ;-)
